System/Directory/Tree.hs

```diff
@@ -353,16 +353,18 @@ buildAtOnce' f p = handleDT n $
      where n = topDir p
 
 
--- using unsafePerformIO to get "lazy" traversal:
+-- using unsafeInterleaveIO to get "lazy" traversal:
 buildLazilyUnsafe' :: Builder a
 buildLazilyUnsafe' f p = handleDT n $
            do isFile <- doesFileExist p
               if isFile
-                 then  File n <$> f p
-                  -- HERE IS THE UNSAFE CODE:
-                 else Dir n . fmap (rec . combine p) <$> getDirsFiles p
-     -- TODO: this should really be unsafeInterleaveIO
-     where rec = unsafePerformIO . buildLazilyUnsafe' f
+                 then File n <$> f p
+                 -- HERE IS THE UNSAFE CODE:
+                 else do
+                   files <- getDirsFiles p
```

Comment:
Note that this line is not behind `unsafeInterleaveIO` on purpose, otherwise the bug in https://github.com/jberryman/directory-tree/issues/9 will cause a runtime exception. Note that that bug was not introduced by this change, it already exists in the original version.